widget/editor.go

```diff
@@ -936,6 +1153,115 @@ func (e *Editor) NumLines() int {
 	return len(e.lines)
 }
 
+// If there's a selection, move to its earliest point and delete the selected
+// text, leaving the editor caret at that point, and clearing the selection.
+//
+// If there's not a selection, don't do anything.
+func (e *Editor) DeleteSelection() {
+	l := e.Selection.Len()
+	if l == 0 {
+		return
+	}
+	e.rr.caret = min(e.Selection.Start.ofs, e.Selection.End.ofs)
+	e.ClearSelection()
+	e.Delete(l)
+}
+
+func (e *Editor) ClearSelection() {
+	e.Selection = Selection{}
+}
+
+func (e *Editor) IsSelected() bool {
```

Comment:
So a positive width means the cursor is at the beginning of the selection, and a negative width means the cursor is at the end of the selection?

Reply:
A problem with this idea is that PageUp/Down move the caret but not the selection.

On the other hand, neither Gmail nor Github do it that way.  (PageUp/Down scroll the screen, but if you then use an arrow key, you scroll back, whether you have a selection or not.)

(Vim does it yet a third way, where starting "visual mode" and then pressing PageUp/Down moves both the cursor and the selection end point; for the Gio editor, it'd be equivalent to treating PageUp/Down as their shifted equivalents if there was a selection.)

So I think I could make that change, at the expense of having PageUp/Down scroll the editor but not move the caret.  Or having PageUp/Down remove the selection.  The latter would be easier to implement, I think, but also I think I prefer the former behavior.

Thoughts?

Reply:
> So a positive width means the cursor is at the beginning of the selection, and a negative width means the cursor is at the end of the selection?

Yes. Just to be clear, I was merely bringing an alternative representation to your attention in case you found it better. I'm fine with (selectionStart, selectionEnd) as well. The main point was that I'd like to assume a selection, and move the zero-width selection special cases as low-level as possible, in order to decrease the risk of forgetting edge cases.

> A problem with this idea is that PageUp/Down move the caret but not the selection.
> 

I'm not sure I see the problem, can you elaborate? I don't mind there be two methods for pageup/down, one for regular pageup/down, and another for shift+pageup/down. In other words, it's fine (and expected) to check for ModShift in high-level code; my hope is to remove checks on SelectionLength/IsSelected in high-level logic.

> On the other hand, neither Gmail nor Github do it that way. (PageUp/Down scroll the screen, but if you then use an arrow key, you scroll back, whether you have a selection or not.)
> 

I think that's a macOS specific behaviour. On my macOS Firefox, the GH comment editor scrolls on pageup/down but don't move the caret. On my Linux Firefox, pageup/down scrolls and moves the caret. I believe, but didn't check, that Windows behave as Linux.

> So I think I could make that change, at the expense of having PageUp/Down scroll the editor but not move the caret. Or having PageUp/Down remove the selection. The latter would be easier to implement, I think, but also I think I prefer the former behavior.
> 

Judging by the kitchen example it seems Gio implements the Linux behaviour, so I'd expect your latter selection behaviour for consistency. Note that I don't mind a future change that changes behaviour according to runtime.GOOS.

Reply:
>  The main point was that I'd like to always assume a selection, and move the zero-width selection special cases as low-level as possible

Sure, I'm fine with that.  Thank you for clarifying, though.

> > A problem with this idea is that PageUp/Down move the caret but not the selection.
>
> I'm not sure I see the problem, can you elaborate?

If selection is necessarily relative to the caret, because selection is defined by the caret + a signed width, then the caret must be either at the beginning or end of the selection.

But in the kitchen app, with this patch currently as-is, PageUp & PageDown move the caret completely outside the selection, without changing the selection.

So if selection is defined as caret + width, then one way or the other the behavior of PageUp/Down needs to change, either moving the caret and thus extending the selection, or by moving the caret and removing the selection, or by scrolling the text but not actually moving the caret.

So having written all that, and rereading your response, I think you're saying that PageUp/Down (that is, `e.movePages(+/-1)`) should move the caret and dismiss the selection, and that I should add another method (in addition to `movePages`) along the lines of `movePagesWithSelection`.  (Or possibly a `select` flag to `movePages`.)

> Judging by the kitchen example it seems Gio implements the Linux behaviour, so I'd expect your latter selection behaviour for consistency.

Okay.

I'm thinking `func (e *Editor) movePages(pages int, select bool)`, which will always move the caret, and optionally create/move/extend the selection.  And also change the calling code to be something like

```
	case key.NamePageUp:
		e.movePages(-1, k.Modifiers == key.ModShift)
	case key.NamePageDown:
		e.movePages(+1, k.Modifiers == key.ModShift)
```

Reply:
> > The main point was that I'd like to always assume a selection, and move the zero-width selection special cases as low-level as possible
> 
> Sure, I'm fine with that. Thank you for clarifying, though.
> 
> > > A problem with this idea is that PageUp/Down move the caret but not the selection.
> > 
> > 
> > I'm not sure I see the problem, can you elaborate?
> 
> If selection is necessarily relative to the caret, because selection is defined by the caret + a signed width, then the caret must be either at the beginning or end of the selection.
> 
> But in the kitchen app, with this patch currently as-is, PageUp & PageDown move the caret completely outside the selection, without changing the selection.
> 
> So if selection is defined as caret + width, then one way or the other the behavior of PageUp/Down needs to change, either moving the caret and thus extending the selection, or by moving the caret and removing the selection, or by scrolling the text but not actually moving the caret.
> 

Indeed, that's one reason I'd like selection-awareness as low-level as possible: so that we're forced to handle all cases, including how to deal with the selection with pageup/down. And by defining caretPos to always be equal selectionStart, we've eliminated one source of inconsistency. (Again, it's up to you whether you like to represent the selection with and absolute selectionEnd or a relative, signed, selectionWidth).

> So having written all that, and rereading your response, I think you're saying that PageUp/Down (that is, `e.movePages(+/-1)`) should move the caret and dismiss the selection, and that I should add another method (in addition to `movePages`) along the lines of `movePagesWithSelection`. (Or possibly a `select` flag to `movePages`.)
> 

Moving the caret and dismissing the selection (or setting selectionEnd to caretPos) SGTM. However,
since you're implementing the macOS behaviour for selection anyway, I'm also ok with you changing the behaviour of pageup/pagedown to just scroll the Editor without moving the caret. If you do, please make the pageup/down behaviour change in a separate commit and place it first in your commit stack (for example by using `git rebase -i origin/main`).

If you choose to keep the current pageup/down behaviour of moving the caret, then adding a flag to movePages sounds best.

> > Judging by the kitchen example it seems Gio implements the Linux behaviour, so I'd expect your latter selection behaviour for consistency.
> 
> Okay.
> 

See above. You're free to implement the behaviour you like best.

> I'm thinking `func (e *Editor) movePages(pages int, select bool)`, which will always move the caret, and optionally create/move/extend the selection. And also change the calling code to be something like
> 
> ```
> 	case key.NamePageUp:
> 		e.movePages(-1, k.Modifiers == key.ModShift)
> 	case key.NamePageDown:
> 		e.movePages(+1, k.Modifiers == key.ModShift)
> ```

Right, SGTM (again assuming you keep the current pageup/down behaviour).

Reply:
A lot of the discussion above was predicated on using a caret+signed length representation for the selection, which I'm not doing, so that part is moot.  I have pushed the selection-aware bits down as far as they can go, and generalized the API to address selection in a universal, consistent manner (I hope, haha).  Please take a look.